### DISCUSSION TITLE
TypeError: Method Map.prototype.set called on incompatible receiver #<Map> at Map.set (<anonymous>)

### DISCUSSION THREAD

**Discussion by Juknum** [mikro-orm/mikro-orm#4519] [General]
> My issue seems to be related to https://github.com/mikro-orm/mikro-orm/discussions/3795 but I was not entirely sure if it was a good idea to continue the discussion or to open a new issue (as it seems to have been fixed with a [PR](https://github.com/mikro-orm/mikro-orm/pull/3796))

**Describe the bug**

Well, I've made a GitHub workflow to runs tests using GitHub actions before merging stuff on my main branch, those tests works fine on my computers (both my mac & windows) but when I do push my changes and when the workflow runs, it fails with the following error :

**Stack trace**
```
    TypeError: Method Map.prototype.set called on incompatible receiver #<Map>
        at Map.set (<anonymous>)

      22 | // So this runs before all tests of the suite
      23 | beforeAll(async () => {
    > 24 |      const moduleFixture: TestingModule = await Test.createTestingModule({
         |                                           ^
      25 |              imports: [AppModule],
      26 |      }).compile();
      27 |
```

**To Reproduce**
Steps to reproduce the behaviour:
1. Clone that [repository](https://github.com/ae-utbm/api/tree/setup-mikrorm) (do not forget submodules)
2. Switch to the `setup-mikrorm` branch _(yes there is a typo in the branch name my bad 😅)_
3. Run `pnpm i` 
4. Execute tests with `pnpm test`
5. See the error (If the error does not show at first, try to play with the `detectOpenHandles` option inside the `jest.config.ts` file)

**Expected behavior**
The workflow should run tests without crashing

**Additional context**
- The [GitHub Action](https://github.com/ae-utbm/api/actions/runs/5513315908/jobs/10051301772) that fails
- The [workflow](https://github.com/ae-utbm/api/blob/setup-mikrorm/.github/workflows/testing.yml) file

> The error seems to fix itself when doing `pnpm jest --clearCache` before running tests, **but it does nothing on the workflow :/**

**Versions**

| Dependency | Version |
| - | - |
| node | 20.4.0 |
| typescript | 5.1.6 |
| mikro-orm | 5.7.12 |
| your-driver | PostgreSQL |
| jest | 29.6.1 |
| ts-jest | 29.1.1 |
| @nestjs/testing | 10.0.5 |
| @mikro-orm/nestjs | 5.2.0 |

**Comment by sevenryze** [comment]
I saw you've set up the global setup feature of jest, try to remove that config line to see if it works again. Your db setup can be moved to `before*` hooks temporarily.

**Comment by B4nan** [comment]
This seems to be jest specific issue, not really something to fix in the ORM (if you think otherwise, repro without jest is welcome). Unfortunately, jest likes to do some weird things under the hood, like patching primitives like `Map`. IMO this is the very same problem as #3795 - that issue wasn't "fixed", the PR you mentioned only added a way to get around it, which could be valid for you as well - by generating the dev metadata upfront via CLI (so outside of jest context, before it does the patching).

As mentioned in the other comment as well as in the linked GH discussion, this is probably connected to the global setup scripts in jest, and you usually don't need that, so another way would be to try to refactor your test setup to not use global setup scripts. It's not great, but you can replace that with a few helper functions and 2-3 lines of code in each of the test cases (and it will end up making the tests more readable anyway).

--- 

I just found this [old issue from jest](https://github.com/jestjs/jest/issues/10199) about this problem, unfortunately completely ignored and closed by a bot, even though it included a reproduction...

**Reply by kallekulp** [reply]
I know this is an old thread, but how would you go with the helpers if each test suite is calling `orm.schema.refreshDatabase()`? When running a test for a whole library the tests are run in parallel, meaning we are getting duplicate table errors. is this expected?

**Reply by B4nan** [reply]
Each test suite should have its own database ideally, otherwise you cant really run them in parallel.

**Comment by Juknum** [comment]
Thanks to both of you, I was struggling so much with this for some days!

This is what I've done (for future readers): 
- Removed the `globalSetup` mention from my Jest settings (that was the main issue)
- Instead used `ts-node` to setup the database with `pnpm ts-node tests/database.setup.ts` (I had to modify the exported function to call itself instead of just being exported)
- Had to modify my workflow to first execute that setup file, then executing tests
